hail/src/main/scala/is/hail/expr/ir/agg/CallStatsAggregator.scala

```diff
@@ -0,0 +1,215 @@
+package is.hail.expr.ir.agg
+
+import is.hail.annotations.{Region, StagedRegionValueBuilder}
+import is.hail.asm4s._
+import is.hail.expr.ir.{EmitMethodBuilder, EmitTriplet}
+import is.hail.expr.types.physical._
+import is.hail.io.{CodecSpec, InputBuffer, OutputBuffer}
+import is.hail.stats.CallStats
+import is.hail.utils._
+
+
+object CallStatsState {
+  val callStatsInternalArrayType = PArray(PInt32Required, required = true)
+  val stateType: PTuple = PTuple(true, callStatsInternalArrayType, callStatsInternalArrayType)
+
+  val resultType = PStruct(
+    "AC" -> PArray(PInt32()),
+    "AF" -> PArray(PFloat64()),
+    "AN" -> PInt32(),
+    "homozygote_count" -> PArray(PInt32()))
+
+  assert(resultType.virtualType == CallStats.schema)
+}
+
+case class CallStatsState(mb: EmitMethodBuilder) extends PointerBasedRVAState {
+  val alleleCountsOffset: Code[Long] = CallStatsState.stateType.fieldOffset(off, 0)
+  val homCountsOffset: Code[Long] = CallStatsState.stateType.fieldOffset(off, 1)
+  val alleleCounts: Code[Long] = CallStatsState.stateType.loadField(off, 0)
+  val homCounts: Code[Long] = CallStatsState.stateType.loadField(off, 1)
+  val nAlleles: Code[Int] = CallStatsState.callStatsInternalArrayType.loadLength(alleleCounts)
+  private val addr = mb.newField[Long]
+
+  // unused but extremely useful for debugging if something goes wrong
+  def dump(tag: String): Code[Unit] = {
+    val i = mb.newField[Int]
+    Code(
+      Code._println(s"at tag $tag"),
+      i := 0,
+      Code.whileLoop(i < nAlleles,
+        Code._println(const("at i=").concat(i.toS).concat(", AC=").concat(alleleCountAtIndex(i, nAlleles).toS).concat(", HOM=").concat(homCountAtIndex(i, nAlleles).toS)),
+        i := i + 1)
+    )
+  }
+
+  def alleleCountAtIndex(idx: Code[Int], length: Code[Int]): Code[Int] =
```

Comment:
another note---instead of creating a new `len` field for each seq op and comb op and passing it in to these functions, you could just store the length inside the state, and just make sure to load the length into the field on load (basically override the existing `load` function to load the length after it's done loading everything else)

Reply:
good idea